cmd/controller-manager/app/controllermanager.go

```diff
@@ -0,0 +1,221 @@
+			// show flags with the final logging configuration.
+			if err := logsapi.ValidateAndApply(o.Logs, utilfeature.DefaultFeatureGate); err != nil {
+				return err
+			}
+
+			// To help debugging, print flags before further processing.
+			cliflag.PrintFlags(cmd.Flags())
+
+			if err := o.Validate(KnownControllers(), ControllersDisabledByDefault.List()); err != nil {
+				return err
+			}
+
+			return Run(ctx, o)
+		},
+		Args: func(cmd *cobra.Command, args []string) error {
+			for _, arg := range args {
+				if len(arg) > 0 {
+					return fmt.Errorf("%q does not take any arguments, got %q", cmd.CommandPath(), args)
+				}
+			}
+			return nil
+		},
+	}
+
+	fs := cmd.Flags()
+	namedFlagSets := o.Flags(KnownControllers(), ControllersDisabledByDefault.List())
+
+	// Add the flag(--kubeconfig) that is added by controller-runtime
+	namedFlagSets.FlagSet("misc").AddGoFlagSet(flag.CommandLine)
+
+	verflag.AddFlags(namedFlagSets.FlagSet("global"))
+	globalflag.AddGlobalFlags(namedFlagSets.FlagSet("global"), cmd.Name(), logs.SkipLoggingConfigurationFlags())
+	for _, f := range namedFlagSets.FlagSets {
+		fs.AddFlagSet(f)
+	}
+
+	cols, _, _ := term.TerminalSize(cmd.OutOrStdout())
+	cliflag.SetUsageAndHelpFunc(cmd, namedFlagSets, cols)
+
+	return cmd
+}
+
+// Run runs the karmada-operator. This should never exit.
+func Run(ctx context.Context, o *options.ControllerManagerOptions) error {
+	// To help debugging, immediately log version
+	// Fixme: define the karmada-operator's own version
+	klog.Infof("Version: %+v", version.Get())
+
+	klog.InfoS("Golang settings", "GOGC", os.Getenv("GOGC"), "GOMAXPROCS", os.Getenv("GOMAXPROCS"), "GOTRACEBACK", os.Getenv("GOTRACEBACK"))
+
+	controllerManager, err := createControllerManager(ctx, o)
+	if err != nil {
+		klog.Errorf("failed to build controller manager: %v", err)
+		return err
+	}
+
+	if err := controllerManager.AddHealthzCheck("ping", healthz.Ping); err != nil {
+		klog.Errorf("failed to add health check endpoint: %v", err)
+		return err
+	}
+
+	controllerCtx, err := CreateControllerContext(o, controllerManager)
+	if err != nil {
+		klog.Errorf("error building controller context: %v", err)
+		return err
+	}
+
+	if err := startUpControllers(controllerCtx, NewControllerInitializers()); err != nil {
+		klog.Errorf("failed to startup controllers: %v", err)
+		return err
+	}
+
+	// blocks until the context is done.
+	if err := controllerManager.Start(ctx); err != nil {
+		klog.Errorf("controller manager exits unexpectedly: %v", err)
+		return err
+	}
+
+	// never reach here
+	return nil
+}
+
+// InitFunc is used to launch a particular controller. It returns a controller
+// that can optionally implement other interfaces so that the controller manager
+// can support the requested features.
+// Any error returned will cause the controller process to `Fatal`
+// The bool indicates whether the controller was enabled.
+type InitFunc func(controllerCtx ControllerContext) (enabled bool, err error)
```

Comment:
These functions could be moved to `core.go`, or putting `pkgcontroller/context/context.go` uniformly.